File: leetcode/weekly-contest/20240324/100258-最高频率的ID.cpp
#include <queue>
#include <unordered_map>
#include <utility>
#include <vector>
using std::pair;
using std::priority_queue;
using std::unordered_map;
using std::vector;

class Solution
{
public:
  vector<long long> mostFrequentIDs(vector<int> &nums, vector<int> &freq)
  {
    int n = nums.size();
    priority_queue<pair<long long, int>> pq;  // {cnt, num}
    unordered_map<int, long long> count;      // num -> cnt
    vector<long long> ans(n);

    for (int i = 0; i < n; ++i) {
      count[nums[i]] += freq[i];
      pq.emplace(count[nums[i]], nums[i]);

      while (true) {
        auto [cnt, num] = pq.top();
        if (cnt != count[num]) {
          pq.pop();
        } else {
          ans[i] = cnt;
          break;
        }
      }
    }
    return ans;
  }
};

File: leetcode/weekly-contest/20240324/100258-最高频率的ID.py
"""
SortedList
https://grantjenks.com/docs/sortedcontainers/sortedlist.html
"""
from collections import Counter
from sortedcontainers import SortedList
from typing import List
from heapq import heappush, heappop

class Solution:
    def mostFrequentIDs(self, nums: List[int], freq: List[int]) -> List[int]:
        cnt = Counter()
        sl = SortedList()
        ans = []
        for x, f in zip(nums, freq):
            sl.discard(cnt[x])  # 多个 cnt[x] 只会移除一个
            cnt[x] += f
            sl.add(cnt[x])
            ans.append(sl[-1])
        return ans

"""

heap & lazy

"""

class Solution:
    def mostFrequentIDs(self, nums: List[int], freq: List[int]) -> List[int]:
        ans = []
        cnt = Counter()
        h = []
        for x, f in zip(nums, freq):
            cnt[x] += f
            heappush(h, (-cnt[x], x))  # 取负号变成最大堆
            while -h[0][0] != cnt[h[0][1]]:  # 堆顶保存的数据已经发生变化
                heappop(h)  # 删除
            ans.append(-h[0][0])
        return ans
